File: C#/Sample1/Sample1/Scenario.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Simphony.Simulation;


namespace Sample1
{
    public class Scenario : DiscreteEventScenario
    {
        DiscreteEventEngine MyEngine;


        //Resource
        private Resource Loader;
        private Resource Spotter;

        private int LoaderCount;
        private int SpotterCount;

        private WaitingFile FileLoader = new WaitingFile("FileLoader", false);
        private WaitingFile FileSpotter = new WaitingFile("FileSpotter", false);


        private NumericStatistic LargeNumericStatistic;
        private NumericStatistic SmallNumericStatistic;
        public IReadOnlyList<double> ObservationForLargeTrucks;
        public IReadOnlyList<double> ObservationForSmallTrucks;
        public double Mean;


        //Duration
        private double LargeTruckHaulingToLoader = 15;
        private double SmallTruckHaulingToLoader = 10;
        private double LargeTruckLoading = 10;
        private double SmallTruckLoading = 5;
        public double LargeTruckHaulingToSpotter = 30;
        public double SmallTruckHaulingToSpotter = 20;
        private double LargeTruckDumping = 5;
        private double SmallTruckDumping = 3;

        #region [- Ctor -]
        public Scenario(DiscreteEventEngine myEngine, int numLoaders, int numSpotters)
        {
            MyEngine = myEngine;
            LoaderCount = numLoaders;
            SpotterCount = numSpotters;
        }
        #endregion



        public override int InitializeScenario()
        {
            Loader = new Resource("Loader", LoaderCount);
            Spotter = new Resource("Spotter", SpotterCount);
            Loader.WaitingFiles.Add(FileLoader);
            Spotter.WaitingFiles.Add(FileSpotter);
            LargeNumericStatistic = new NumericStatistic("LargeCT", false);
            SmallNumericStatistic = new NumericStatistic("SmallCT", false);

            return 1;
        }

        public override double InitializeRun(int runIndex)
        {
            Loader.InitializeRun(runIndex);
            Spotter.InitializeRun(runIndex);
            FileLoader.InitializeRun(runIndex);
            FileSpotter.InitializeRun(runIndex);
            LargeNumericStatistic.InitializeRun(runIndex);
            SmallNumericStatistic.InitializeRun(runIndex);

            for (int i = 0; i < 4; i++)
            {
                Truck MyTruck = new Truck();

                if (i % 2 == 0)
                {
                    MyTruck.Type = "Large";
                    MyEngine.ScheduleEvent(MyTruck, RequestLoader, LargeTruckHaulingToLoader);
                }

                else
                {
                    MyTruck.Type = "Small";
                    MyEngine.ScheduleEvent(MyTruck, RequestLoader, SmallTruckHaulingToLoader);
                }

                MyTruck.StartTime = MyEngine.TimeNow;

            }


            return double.PositiveInfinity;
        }

        private void RequestLoader(Truck myTruck)
        {
            MyEngine.RequestResource(myTruck, Loader, 1, Loading, FileLoader);
        }

        private void Loading(Truck myTruck)
        {
            if (myTruck.Type == "Large")
            {
                MyEngine.ScheduleEvent(myTruck, ReleaseLoaderAndHauling, LargeTruckLoading);
            }

            else
            {
                MyEngine.ScheduleEvent(myTruck, ReleaseLoaderAndHauling, SmallTruckLoading);
            }
        }

        private void ReleaseLoaderAndHauling(Truck myTruck)
        {
            MyEngine.ReleaseResource(myTruck, Loader, 1);


            //Hauling
            if (myTruck.Type == "Large")
            {
                MyEngine.ScheduleEvent(myTruck, RequestSpotter, LargeTruckHaulingToSpotter);
            }

            else
            {
                MyEngine.ScheduleEvent(myTruck, RequestSpotter, SmallTruckHaulingToSpotter);
            }
        }

        private void RequestSpotter(Truck myTruck)
        {
            MyEngine.RequestResource(myTruck, Spotter, 1, Dumping, FileSpotter);
        }


        private void Dumping(Truck myTruck)
        {
            if (myTruck.Type == "Large")
            {
                MyEngine.ScheduleEvent(myTruck, ReleaseSpotter, LargeTruckDumping);
            }

            else
            {
                MyEngine.ScheduleEvent(myTruck, ReleaseSpotter, SmallTruckDumping);
            }
        }

        private void ReleaseSpotter(Truck myTruck)
        {
            MyEngine.ReleaseResource(myTruck, Spotter, 1);

            if (myTruck.Type == "Large")
            {
                MyEngine.CollectStatistic(LargeNumericStatistic, MyEngine.TimeNow - myTruck.StartTime);
            }

            else
            {
                MyEngine.CollectStatistic(SmallNumericStatistic, MyEngine.TimeNow - myTruck.StartTime);
            }

        }


        public override void FinalizeRun(int runIndex)
        {


        }

        public override void FinalizeScenario()
        {
            ObservationForLargeTrucks = LargeNumericStatistic.Times;
            ObservationForSmallTrucks = SmallNumericStatistic.Times;

            Mean = (LargeNumericStatistic.Sum + SmallNumericStatistic.Sum) / (LargeNumericStatistic.Count + SmallNumericStatistic.Count);
        }


    }
}

File: VB/Sample1/Sample1/Scenario.vb
Imports Simphony.Simulation
Imports Simphony.Mathematics

Public Class Scenario
    Inherits DiscreteEventScenario

    Private ReadOnly MyEngine As DiscreteEventEngine

    Private RequestResourceEvent As New Action(Of Truck)(AddressOf RequestResource)
    Private ResourceIsCapturedEvent As New Action(Of Truck)(AddressOf ResourceIsCaptured)
    Private ReleaseResourceEvent As New Action(Of Truck)(AddressOf ReleaseResource)

    'Resource
    Private LargeLoaders, SmallLoaders As LoaderResource
    Private LargeLoaderCount, SmallLoaderCount As Integer

    Private ReadOnly MyFile As New WaitingFile("MyFile")                '****************

    Private MyNumericStatistic As NumericStatistic
    Private MyForm As Form1


    Private ReadOnly TotalSimulationTime As Double = 2000


    'Ctor
    Public Sub New(ByVal Engine1 As DiscreteEventEngine, ByVal NumLargeLoaders As Integer, ByVal NumSmallLoaders As Integer, ByVal Form As Form1)
        MyEngine = Engine1
        LargeLoaderCount = NumLargeLoaders
        SmallLoaderCount = NumSmallLoaders
        MyForm = Form
    End Sub


    Public Overrides Function InitializeScenario() As Integer
        LargeLoaders = New LoaderResource("LargeLoaders", LargeLoaderCount, "Large")
        SmallLoaders = New LoaderResource("SmallLoaders", SmallLoaderCount, "Small")
        LargeLoaders.WaitingFiles.Add(MyFile)
        SmallLoaders.WaitingFiles.Add(MyFile)

        MyNumericStatistic = New NumericStatistic("CT", False)

        Return 1
    End Function

    Public Overrides Function InitializeRun(runIndex As Integer) As Double
        LargeLoaders.InitializeRun(runIndex)                                '****************
        SmallLoaders.InitializeRun(runIndex)                                '****************
        MyFile.InitializeRun(runIndex)                                      '****************
        MyNumericStatistic.InitializeRun(runIndex)

        For i As Integer = 1 To 10
            Dim MyTruck As New Truck
            If i Mod 2 = 0 Then
                MyTruck.Type = "Small"
                MyTruck.Size = 10
            Else
                MyTruck.Type = "Large"
                MyTruck.Size = 20
            End If

            MyEngine.ScheduleEvent(MyTruck, RequestResourceEvent, 10)

        Next

        Return Double.PositiveInfinity
    End Function

    Private Sub RequestResource(MyTruck As Truck)

        MyTruck.StartTime = MyEngine.TimeNow

        If MyTruck.Type = "Large" Then
            MyEngine.RequestResource(MyTruck, LargeLoaders, 1, ResourceIsCapturedEvent, MyFile)
        Else
            MyEngine.RequestResource(MyTruck, SmallLoaders, 1, ResourceIsCapturedEvent, MyFile)
        End If

    End Sub

    Private Sub ResourceIsCaptured(MyTruck As Truck)
        MyEngine.ScheduleEvent(MyTruck, ReleaseResourceEvent, 20)
    End Sub

    Private Sub ReleaseResource(MyTruck As Truck)

        If MyTruck.Type = "Large" Then
            MyEngine.ReleaseResource(MyTruck, LargeLoaders, 1)
        Else
            MyEngine.ReleaseResource(MyTruck, SmallLoaders, 1)
        End If

        MyEngine.CollectStatistic(MyNumericStatistic, MyEngine.TimeNow - MyTruck.StartTime)

    End Sub

    Public Overrides Sub FinalizeRun(runIndex As Integer)

    End Sub

    Public Overrides Sub FinalizeScenario()

        Dim Mean As Double = MyNumericStatistic.Mean
        Dim StandardDeviation As Double = MyNumericStatistic.StandardDeviation

        MyForm.ListBox1.Items.Add("Mean= " & Mean.ToString & ", Standard Deviation= " & StandardDeviation.ToString)
    End Sub


End Class
